File: plugins/plugin-solana/typescript/providers/wallet.ts
import type { IAgentRuntime, Memory, Provider, ProviderResult, State } from "@elizaos/core";
import { logger } from "@elizaos/core";
import BigNumber from "../bn";
import { SOLANA_WALLET_DATA_CACHE_KEY } from "../constants";
import { getWalletKey } from "../keypairUtils";
import type { WalletPortfolio } from "../types";

/**
 * Wallet provider for Solana.
 * @param {IAgentRuntime} runtime - The agent runtime.
 * @param {Memory} _message - The memory message.
 * @param {State} [state] - Optional state parameter.
 * @returns {Promise<ProviderResult>} The result of the wallet provider.
 */
export const walletProvider: Provider = {
  name: "solana-wallet",
  description: "your solana wallet information",
  // it's not slow we always have this data
  // but we don't always need this data, let's free up the context
  dynamic: true,
  get: async (runtime: IAgentRuntime, _message: Memory, state: State): Promise<ProviderResult> => {
    try {
      const portfolioCache = await runtime.getCache<WalletPortfolio>(SOLANA_WALLET_DATA_CACHE_KEY);
      if (!portfolioCache) {
        logger.info("solana::wallet provider - portfolioCache is not ready");
        return { data: {}, values: {}, text: "" };
      }

      const solanaService = runtime.getService("solana");
      let pubkeyStr = "";

      const { publicKey } = await getWalletKey(runtime, false);

      if (solanaService && publicKey) {
        pubkeyStr = ` (${publicKey.toBase58()})`;
      }

      const portfolio = portfolioCache;
      const agentName = state.agentName ?? runtime.character.name ?? "The agent";
      const totalSol = portfolio.totalSol ?? "0";

      // Values that can be injected into templates
      const values: Record<string, string> = {
        total_usd: new BigNumber(portfolio.totalUsd).toFixed(2),
        total_sol: totalSol,
      };

      // Add token balances to values
      portfolio.items.forEach((item, index) => {
        if (new BigNumber(item.uiAmount).isGreaterThan(0)) {
          values[`token_${index}_name`] = item.name;
          values[`token_${index}_symbol`] = item.symbol;
          values[`token_${index}_amount`] = new BigNumber(item.uiAmount).toFixed(6);
          values[`token_${index}_usd`] = new BigNumber(item.valueUsd).toFixed(2);
          values[`token_${index}_sol`] = item.valueSol ?? "0";
        }
      });

      // Add market prices to values
      if (portfolio.prices) {
        values.sol_price = new BigNumber(portfolio.prices.solana.usd).toFixed(2);
        values.btc_price = new BigNumber(portfolio.prices.bitcoin.usd).toFixed(2);
        values.eth_price = new BigNumber(portfolio.prices.ethereum.usd).toFixed(2);
      }

      // Format the text output
      let text = `\n\n${agentName}'s Main Solana Wallet${pubkeyStr}\n`;
      text += `Total Value: $${values.total_usd} (${values.total_sol} SOL)\n\n`;

      // Token Balances
      text += "Token Balances:\n";
      const nonZeroItems = portfolio.items.filter((item) =>
        new BigNumber(item.uiAmount).isGreaterThan(0)
      );

      if (nonZeroItems.length === 0) {
        text += "No tokens found with non-zero balance\n";
      } else {
        for (const item of nonZeroItems) {
          const valueUsd = new BigNumber(item.valueUsd).toFixed(2);
          const valueSol = item.valueSol ?? "0";
          text += `${item.name} (${item.symbol}): ${new BigNumber(item.uiAmount).toFixed(
            6
          )} ($${valueUsd} | ${valueSol} SOL)\n`;
        }
      }

      // Market Prices
      if (portfolio.prices) {
        text += "\nMarket Prices:\n";
        text += `SOL: $${values.sol_price}\n`;
        text += `BTC: $${values.btc_price}\n`;
        text += `ETH: $${values.eth_price}\n`;
      }

      // Portfolio data for the ProviderResult
      const data = {
        totalUsd: portfolio.totalUsd,
        totalSol: portfolio.totalSol,
        items: portfolio.items,
        prices: portfolio.prices,
        lastUpdated: portfolio.lastUpdated,
      };

      return {
        data,
        values,
        text,
      };
    } catch (error) {
      logger.error(
        `Error in Solana wallet provider: ${error instanceof Error ? error.message : String(error)}`
      );
      return { data: {}, values: {}, text: "" };
    }
  },
};

File: plugins/plugin-solana/rust/src/providers/wallet.rs
#![allow(missing_docs)]

use rust_decimal::Decimal;
use std::collections::HashMap;
use tracing::{error, info};

use crate::client::SolanaClient;
use crate::types::{PortfolioItem, WalletPortfolio};
use crate::WRAPPED_SOL_MINT;

#[derive(Debug, Clone)]
pub struct WalletProviderResult {
    pub data: WalletPortfolio,
    pub values: HashMap<String, String>,
    pub text: String,
}

pub struct WalletProvider;

impl WalletProvider {
    pub const NAME: &'static str = "solana-wallet";
    pub const DESCRIPTION: &'static str = "your solana wallet information";
    pub const DYNAMIC: bool = true;

    pub async fn get(
        client: &SolanaClient,
        agent_name: Option<&str>,
    ) -> Result<WalletProviderResult, String> {
        let agent_name = agent_name.unwrap_or("The agent");
        let pubkey = client.public_key();
        let pubkey_str = pubkey.to_string();

        info!("Fetching wallet portfolio for {}", pubkey_str);

        let sol_balance = match client.get_sol_balance().await {
            Ok(b) => b,
            Err(e) => {
                error!("Failed to get SOL balance: {}", e);
                return Err(format!("Failed to get SOL balance: {}", e));
            }
        };

        let token_accounts = match client.get_token_accounts().await {
            Ok(ta) => ta,
            Err(e) => {
                error!("Failed to get token accounts: {}", e);
                vec![]
            }
        };

        let sol_price = match client
            .get_token_prices(&[WRAPPED_SOL_MINT.to_string()])
            .await
        {
            Ok(prices) => prices.get(WRAPPED_SOL_MINT).copied(),
            Err(_) => None,
        };

        let sol_value_usd = if let Some(price) = sol_price {
            sol_balance * Decimal::from_f64_retain(price).unwrap_or(Decimal::ZERO)
        } else {
            Decimal::ZERO
        };

        let mut items: Vec<PortfolioItem> = vec![];

        items.push(PortfolioItem {
            name: "Solana".to_string(),
            symbol: "SOL".to_string(),
            address: WRAPPED_SOL_MINT.to_string(),
            decimals: 9,
            balance: (sol_balance * Decimal::from(1_000_000_000u64)).to_string(),
            ui_amount: sol_balance.to_string(),
            price_usd: sol_price
                .map(|p| p.to_string())
                .unwrap_or_else(|| "0".to_string()),
            value_usd: sol_value_usd.to_string(),
            value_sol: Some(sol_balance.to_string()),
        });

        for account in token_accounts {
            if account.ui_amount > Decimal::ZERO {
                items.push(PortfolioItem {
                    name: account.mint.clone(),
                    symbol: "TOKEN".to_string(),
                    address: account.mint,
                    decimals: account.decimals,
                    balance: account.amount.clone(),
                    ui_amount: account.ui_amount.to_string(),
                    price_usd: "0".to_string(),
                    value_usd: "0".to_string(),
                    value_sol: None,
                });
            }
        }

        let total_usd = sol_value_usd;

        let portfolio = WalletPortfolio {
            total_usd: total_usd.to_string(),
            total_sol: Some(sol_balance.to_string()),
            items: items.clone(),
            prices: None,
            last_updated: Some(
                std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .map(|d| d.as_millis() as u64)
                    .unwrap_or(0),
            ),
        };

        let mut values: HashMap<String, String> = HashMap::new();
        values.insert("total_usd".to_string(), total_usd.to_string());
        values.insert("total_sol".to_string(), sol_balance.to_string());

        if let Some(price) = sol_price {
            values.insert("sol_price".to_string(), price.to_string());
        }

        for (idx, item) in items.iter().enumerate() {
            values.insert(format!("token_{}_name", idx), item.name.clone());
            values.insert(format!("token_{}_symbol", idx), item.symbol.clone());
            values.insert(format!("token_{}_amount", idx), item.ui_amount.clone());
            values.insert(format!("token_{}_usd", idx), item.value_usd.clone());
            if let Some(ref sol_val) = item.value_sol {
                values.insert(format!("token_{}_sol", idx), sol_val.clone());
            }
        }

        let mut text = format!("\n\n{}'s Main Solana Wallet ({})\n", agent_name, pubkey_str);
        text.push_str(&format!(
            "Total Value: ${} ({} SOL)\n\n",
            total_usd, sol_balance
        ));
        text.push_str("Token Balances:\n");

        if items.is_empty() {
            text.push_str("No tokens found with non-zero balance\n");
        } else {
            for item in &items {
                let sol_str = item
                    .value_sol
                    .as_ref()
                    .map(|s| format!(" | {} SOL", s))
                    .unwrap_or_default();
                text.push_str(&format!(
                    "{} ({}): {} (${}{})\n",
                    item.name, item.symbol, item.ui_amount, item.value_usd, sol_str
                ));
            }
        }

        if let Some(price) = sol_price {
            text.push_str(&format!("\nMarket Prices:\nSOL: ${:.2}\n", price));
        }

        Ok(WalletProviderResult {
            data: portfolio,
            values,
            text,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[allow(clippy::assertions_on_constants)]
    fn test_provider_metadata() {
        assert_eq!(WalletProvider::NAME, "solana-wallet");
        assert!(!WalletProvider::DESCRIPTION.is_empty());
        assert!(WalletProvider::DYNAMIC);
    }
}
